proto/symge/symge-demo.cpp

```diff
@@ -0,0 +1,387 @@
+// bloody CMake
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <cassert>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <random>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+#include <utility>
+
+#include <util/optional.hpp>
+
+#include "symbolic.hpp"
+#include "msparse.hpp"
+
+using namespace sym;
+
+// Identifier name picking helper routines
+
+void join_impl(std::ostream& ss, const std::string& sep) {}
+
+template <typename Head, typename... Args>
+void join_impl(std::ostream& ss, const std::string& sep, Head&& head, Args&&... tail) {
+    if (sizeof...(tail)==0)
+        ss << std::forward<Head>(head);
+    else
+        join_impl(ss << std::forward<Head>(head) << sep, sep, std::forward<Args>(tail)...);
+}
+
+template <typename... Args>
+std::string join(const std::string& sep, Args&&... items) {
+    std::stringstream ss;
+    join_impl(ss, sep, std::forward<Args>(items)...);
+    return ss.str();
+}
+
+class id_maker {
+private:
+    std::unordered_set<std::string> ids;
+
+public:
+    static std::string next_id(std::string s) {
+        unsigned l = s.size();
+        if (l==0) return "a";
+
+        unsigned i = l-1;
+
+        char l0='a', l1='z';
+        char u0='A', u1='Z';
+        char d0='0', d1='9';
+        for (;;) {
+            char& c = s[i];
+            if ((c>=l0 && c<l1) || (c>=u0 && c<u1) || (c>=d0 && c<d1)) {
+                ++c;
+                return s;
+            }
+            if (c==l1) c=l0;
+            if (c==u1) c=u0;
+            if (c==d1) c=d0;
+            if (i==0) break;
+            --i;
+        }
+
+        // prepend a character based on class of first
+        if (s[0]==u0) return u0+s;
+        if (s[0]==d0) return d0+s;
+        return l0+s;
+    }
+
+    template <typename... Args>
+    std::string operator()(Args&&... elements) {
+        std::string name = join("",std::forward<Args>(elements)...);
+        if (name.empty()) name = "a";
+
+        while (ids.count(name)) {
+            name = next_id(name);
+        }
+        ids.insert(name);
+        return name;
+    }
+
+    void reserve(std::string name) {
+        ids.insert(std::move(name));
+    }
+};
+
+// Output helper functions
+
+template <typename X>
+std::ostream& operator<<(std::ostream& o, const optional<X>& x) {
+    return x? o << *x: o << "nothing";
+}
+
+template <typename X>
+std::ostream& operator<<(std::ostream& o, const msparse::matrix<X>& m) {
+    for (unsigned r = 0; r<m.nrow(); ++r) {
+        o << '|';
+        for (unsigned c = 0; c<m.ncol(); ++c) {
+            if (c==m.augcol()) o << " | ";
+            o << std::setw(12) << m[r][c];
+        }
+        o << " |\n";
+    }
+    return o;
+}
+
+template <typename Sep, typename V>
+struct sepval_t {
```

Comment:
I think we have a convention of using `_type`.
This should go in the coding standards, not that anybody reads them.